src/query/somepath.go

```diff
@@ -4,60 +4,88 @@ import "fmt"
+	}
+	for _, l1 := range expandAllTargets(graph, from) {
+		for _, l2 := range expandAllTargets(graph, to) {
+			if path := s.SomePath(graph.TargetOrDie(l1), graph.TargetOrDie(l2)); len(path) != 0 {
+				fmt.Println("Found path:")
+				for _, l := range filterPath(path) {
+					fmt.Printf("  %s\n", l)
+				}
 				return
 			}
 		}
-		fmt.Printf("Couldn't find any dependency path between %s and %s\n", label1, label2)
-	} else {
-		querySomePath1(graph, graph.TargetOrDie(label1), label2, true)
 	}
+	log.Fatalf("Couldn't find any dependency path between those targets")
 }
 
-func querySomePath1(graph *core.BuildGraph, target1 *core.BuildTarget, label2 core.BuildLabel, print bool) bool {
-	// Now we do the same for label2.
-	if label2.IsAllTargets() {
-		for _, target2 := range graph.PackageOrDie(label2).AllTargets() {
-			if querySomePath2(graph, target1, target2, false) {
-				return true
+// expandAllTargets expands any :all labels in the given set.
+func expandAllTargets(graph *core.BuildGraph, labels []core.BuildLabel) []core.BuildLabel {
+	ret := make([]core.BuildLabel, 0, len(labels))
+	for _, l := range labels {
+		if l.IsAllTargets() {
+			for _, t := range graph.PackageOrDie(l).AllTargets() {
+				ret = append(ret, t.Label)
 			}
+		} else {
+			ret = append(ret, l)
 		}
-		return false
 	}
-	return querySomePath2(graph, target1, graph.TargetOrDie(label2), print)
+	return ret
 }
 
-func querySomePath2(graph *core.BuildGraph, target1, target2 *core.BuildTarget, print bool) bool {
-	if !printSomePath(graph, target1, target2) && !printSomePath(graph, target2, target1) {
-		if print {
-			fmt.Printf("Couldn't find any dependency path between %s and %s\n", target1.Label, target2.Label)
-		}
-		return false
+type somepath struct {
+	memo map[core.BuildLabel]map[core.BuildLabel]struct{}
+}
+
+func (s *somepath) SomePath(target1, target2 *core.BuildTarget) []core.BuildLabel {
+	// Have to try this both ways around since we don't know which is a dependency of the other.
+	if path := s.somePath(target1, target2); len(path) != 0 {
+		return path
 	}
-	return true
+	return s.somePath(target2, target1)
 }
 
-// This is just a simple DFS through the graph.
-func printSomePath(graph *core.BuildGraph, target1, target2 *core.BuildTarget) bool {
-	if target1 == target2 {
-		fmt.Printf("Found path:\n  %s\n", target1.Label)
-		return true
+func (s *somepath) somePath(target1, target2 *core.BuildTarget) []core.BuildLabel {
+	m, present := s.memo[target2.Label]
+	if !present {
+		m = map[core.BuildLabel]struct{}{}
+		s.memo[target2.Label] = m
 	}
-	for _, target := range graph.ReverseDependencies(target2) {
-		if printSomePath(graph, target1, target) {
-			if target2.Parent(graph) != target {
-				fmt.Printf("  %s\n", target2.Label)
-			}
-			return true
+	return s.somePath2(target1, target2, m)
+}
+
+func (s *somepath) somePath2(target1, target2 *core.BuildTarget, seen map[core.BuildLabel]struct{}) []core.BuildLabel {
```

Comment:
This doesn't actually use `s` so maybe we can just make this `func somePath(...)`